### ISSUE TITLE
Prevent user from interacting with iframe

### ISSUE THREAD

**Issue by stvstnfrd** [Mindtoeye/XBlockStattutor#14]
## Error

> ERROR: Error parsing xmlFile
> /xblock/resource/stattutor/public/problem_files/m2_body_image/body_image.brd
## Explanation

Loading any XBlock instance in the LMS results in an undismissable
modal: "Error parsing xmlFile". I can't close the modal and I'm
prevented from interacting with the iframe due to a translucent overlay.
This happened for both modules I attempted.
## Reproduction (in devstack)
- Add a stattutor XBlock to a course via CMS.
- View XBlock in LMS.
- See modal upon page load.
## Screenshots

![screen shot 2016-07-15 at 13 16 08](https://cloud.githubusercontent.com/assets/7300975/16890674/94125766-4aa5-11e6-95cb-c55913426fb5.png)

**Comment by Ringenberg** [comment]
I am unable to reproduce this problem in my devstack. Looking at some of the CTAT code, that error message is likely due to not being able to find the specified required file.

**Comment by stvstnfrd** [comment]
Ah, I see what the issue was here. It looks to be a problem with content
types, expected vs. actual.

The screenshots I had attached are from an OpenEdX "Devstack" [1], the
canonical developer environment for the edX platform. In this
environment, `.brd` files appear to not be served by the webserver.
Whereas in a separate development environment of mine, the mimetype was
incorrectly assumed to the HTML (the files are really XML), though the
request still manages to succeed.

We can't assume the behavior of the latter case, so we'll want to ensure
this works under default configurations. Since these files are already
XML under-the-hood, the most straight-forward approach I see would be to
simply rename and serve the `.brd` files as `.xml`; we could just add a
suffix like `survey_1.brd.xml`. This should allow any webserver to
easily detect the mimetype directly from the file extension, as
expected.

Any perceived complications with this approach?

The other issue we need to address is that we can't allow
filesystem access that depends on user-provided input, ala selecting the
BRD input file; doing so opens an attack surface that could risk
exposing external files. We'll need an explicit whitelist to check against.
Further, we shouldn't be accessing the FS repeatedly at run-time (e.g.,
when saving via Studio). Instead, we should query the FS _once_ at
"compile" time.

I envision something like:

``` python
# Top-level, imports, etc.
PROBLEM_DIRECTORIES = {
    # Create a dictionary mapping base names to longer paths
    directory: 'public/problem_files/{directory}'.format(
        directory=directory,
    )
    for directory in pkg_resources.resource_listdir(
        __name__,
        'public/problem_files/'
    )
    if pkg_resources.resource_isdir(
        __name__,
        'public/problem_files/{directory}'.format(
            directory,
        )
    )
}

# Later on in studio_view
options = [
    # Build the markup
    "<option value="{short}"{long}>public/problem_files/{short}</option>".format(
        short=short,
        long=long,
        # add your "selected" logic here
    )
    for short, long in PROBLEM_DIRECTORIES
]
options.sort()
options = ''.join(options)

# Later on in studio_save
problem = data.get('problem_file')
if problem in PROBLEM_DIRECTORIES:
    # User has selected a supported file:
    # proceed
else:
    # this should only happen for malicious users, but we still need to
    # catch it
```

This is the major outstanding issue I see, pending the previously opened pull request.
Let me know what you think.

[1] https://openedx.atlassian.net/wiki/display/OpenOPS/Running+Devstack

**Comment by Ringenberg** [comment]
I am using a freshly installed version of the dogwood fullstack and I am still unable to reproduce this problem. 

As for renaming the files, we would have to change all of the references to that file and change the bit of code at https://github.com/CMUCTAT/XBlockStattutor/blob/master/stattutor/stattutor.py#L224 that locates the various files inside the module directories as it currently expects to find a single .xml and a single .brd file.

**Comment by stvstnfrd** [comment]
We aren't running Dogwood, so it's possible that's got something to do
with it. For reference, here's the Vagrantfile [1] we use to provision
our development instances. If you provision your devstack with that,
you'll be working in the same environment as we are.

If you give it a go with that Vagrantfile, I'll also have another one of
our developers try running in their stack and we can all come up with a
plan if we're still seeing conflicting results.

> change the bit of code at

Per the above feedback (see the code samples), you'll already need to
rework the way your `studio_view` and `studio_edit` handle those file
names, since you can't access the filesystem on each save. I can't
imagine changing your string-matching from `.brd` to `.brd.xml` would be
particularly difficult.

> we would have to change all of the references to that file

Unless I'm missing something, should this be as simple as search/replace
across this repo?
I'm only seeing two references to a '.brd' extension in the Python code.

Or is there some code and/or configuration which I haven't seen yet?

[1] https://raw.githubusercontent.com/Stanford-Online/configuration/master/vagrant/release/devstack/Vagrantfile

**Comment by stvstnfrd** [comment]
Another of our team members installed this XBlock into their (post-Cedar, pre-Dogwood, Stanford-master) devstack. Similarly, they were unable to dismiss the modal.

We you able to attempt with the Stanford devstack on your end?

**Comment by Ringenberg** [comment]
A potential fix might be to add .brd as a valid mimetype in OpenEdX by adding it explicitly in /edx/app/edxapp/edx-platform/[cms,lms]/startup.py in add_mimetypes()

**Comment by stvstnfrd** [comment]
Thanks @Ringenberg . 

To avoid hard-coding core-platform code to support a third-party extension, I think we can make this work and generalize it for future work.

I'll write up an abstraction that allows us to selectively define custom mimetype mappings on a per-machine basis. That way, we can add/remove the extensions on an environment by environment basis, by generalizing this for broader use.

I've got the code stubbed out, still needs some testing. I probably won't finish tonight, but I'll have something for Monday.

Which mimetype should be paired with the `.brd` extension?

**Comment by Mindtoeye** [comment]
Steven, before you put too much effort into this, please be aware that this XBlock was specifically designed for the stattutor project. All future CTAT tutors should use our generalized XBlock tutor driver at: https://github.com/CMUCTAT/XBlockCTAT

**Comment by Ringenberg** [comment]
text/xml

**Comment by stvstnfrd** [comment]
> Steven, before you put too much effort into this, please be aware that this XBlock was specifically designed for the stattutor project. All future CTAT tutors should use our generalized XBlock tutor driver at: https://github.com/CMUCTAT/XBlockCTAT

@Mindtoeye Now I'm more confused than ever.
CC: @Ringenberg

TL;DR: This _needs_ fixed in one way or another.
There are various approaches to solving this, but I haven't been given
enough information.

To make sure we're on the same page:
- The stattutor XBlock, in its current form, does _not_ work inside a
  devstack running Stanford's codebase.
- If code doesn't work in our development environment, we can _not_
  install it to production.
- Your team _must_ perform testing within the Stanford environment.
  - In this context, edX Named Releases are irrelevant. If you,
    separately, also want the XBlock to work at edx.org, then you
    would also want to test using _their_ current codebase (they don't
    even use Named Releases).
    Right now, however, I'm only concerned with this working w/ our
    environment.

While we definitely need to have a separate conversation about this
other XBlock [1], I'm not sure how that moves us beyond the current
impasse.

Due to this XBlock's custom use of mimetypes [2], it's not surprising
that we need to manually support it; regardless, we do.
I see two possible solutions here:
- My original suggestion: we add a `.xml` extension to BRD files. The
  major benefits here are:
  - Universal client/server support. You'd be hard pressed to find any
    webserver, browser, or computer that doesn't know how to handle XML.
  - Requires no modification to the OpenEdX server.
    - This is particularly important because you risk needing to modify
      every other server on which you wish to deploy this.
  - Avoids the explicit dependency support between platform and
    extension (XBlock), which I really prefer.
  - I have already tested this and it works. In fact, doing so is the
    only reason I was able to test the rest of the XBlock at all.
- Per your team's suggestion, I offered a workaround which would allow
  us to make custom mimetype associations
  - Avoids renaming any files, but requires modification of core
    platform code on any systems that need to serve this code.
  - Frankly, I _really_ don't like this approach; I offered to take it
    on, mostly as a gesture of good will. But if you're having second
    thoughts now, then I am too.

When I originally proposed the former solution, the push back was "we
would need to update all the references". As I replied, and to which I
have not heard a response, what are those references and/or
dependencies? The references in the code base base appear trivial to
modify. So acting on the limited information which I have [1], I have to
assume that you're referring to external resources of some kind?
Please, what are they?? What do they look like? How many are there?

I can imagine a number of possibilities, but what really matters is your
actual use case. Are there external references that haven't been revealed to me?

If not, please let me know what the complication is.
If so, please tell me as much because there are a number of
approaches we could use to solve this problem.

But again, and I can't emphasize this enough, I can't help you find the
optimal solution if you don't state your assumptions for me. You just now
mentioned that this XBlock is related to a more generalized one, but
that's the first I've ever heard of it. Other comments have mentioned
some form of "we need this because we're using it elsewhere", without
any context as to how that relates to the code that we're reviewing/installing
here and now.

I really want to make it clear that this isn't at all meant to be a "my way
or the highway" thing; it's just that I can _only_ act on the intel I've
been provided; my proposals and problem solving reflect this.

"works on my local" and "we can't do that" aren't enough.

With the information I have been provided, option #1 is objectively more
resilient. If you have information to the contrary, please share it.

[1] Please remember that I know absolutely _nothing_  about the `stattutor`
    XBlock, or its intended use, beyond this code repository. Further,
    this is literally the first I've ever heard of the `CTAT` XBlock.
[2] I wouldn't expect most web servers to know that `.brd` is really an
    XML file. In fact, I hadn't even heard of it until this project.
    FWIW: The only significant hits I see on Google mention a binary
    format for PCB designs..

**Comment by Ringenberg** [comment]
I just merged a potential fix for this issue. The contents of the brd file are now in one of the configuration variables, therefor django no longer needs to be able to serve the brd files.  I have tested it in the xblock-sdk and a dogwood fullstack and it at least does not introduce any new issues in those environments.

**Comment by Ringenberg** [comment]
This fix seems to work on Lagunita as well.
